### ISSUE TITLE
After upgrading Unity to 5.9.4, Compiler error on unreferenced assembly

### ISSUE THREAD

**Issue by danrien** [Particular/NServiceBus.Unity#100]
Steps to reproduce:

1. Update Unity to 5.9.4
2. Compile project

Expected Results:
Project compiles

Actual Results:
Compiler error occurs:

```
  SqlEndpointConfigurationBuilder.cs(62, 58): [CS0012] The type 'IUnityContainer' is defined in an assembly that is not referenced. You must add a reference to assembly 'Unity.Abstractions, Version=3.3.0.0, Culture=neutral, PublicKeyToken=6d32ff45e0ccc69f'.
```

**Comment by ramonsmits** [comment]
@danrien This seems to be a breaking change introduced by the Unity team.

What NServiceBus packages + versions are you currently using?

**Comment by WilliamBZA** [comment]
I've confirmed that this is a breaking change from the Unity team. The 5.9.4 release should have been released as a major and not a minor as the API surface has changed.

We'll have to release a new major of NServiceBus.Unity to address this.

**Comment by WilliamBZA** [comment]
Thanks @danrien. We'll schedule this to be fixed, but for now is it possible to downgrade to version `5.8.*` of Unity?

**Comment by danrien** [comment]
@WilliamBZA yes, we're currently running with the `5.8.*` version of Unity. For what it's worth, I've filed an issue over at Unity asking for some documentation on their versioning strategy - https://github.com/unitycontainer/unity/issues/279

**Comment by ENikS** [comment]
Unity 5.9+ is controlled and references [Unity.Abstractions 4.0+](https://github.com/unitycontainer/container/blob/v5.x/package.props) 
If you update the one you should update the other as well.

**Comment by bording** [comment]
@ENikS The problem is that you changed your strong-naming key between Unity.Abstractions 3.3.0 and 4.0,0, so it's impossible for the type system to unify types. They are completely different types, so that is a binary-breaking change, and requires every assembly referencing it to be recompiled.

If you had kept the key the same, then the 5.8 -> 5.9 update would have been transparent, assuming there were no breaking API changes.

**Comment by ENikS** [comment]
You are right about the key. Unfortunately there were no way around it.
Changing from v5.8 to v5.9 would require recompile anyway. 

Whole Unity API has changed dramatically with new Unity.Abstractions v4.0.0. 
Entire categories of classes have been deprecated.

**Comment by bording** [comment]
> You are right about the key. Unfortunately there were no way around it. 
>Changing from v5.8 to v5.9 would require recompile anyway. 

>Whole Unity API has changed dramatically with new Unity.Abstractions v4.0.0. Entire categories of classes have been deprecated.

Then you should have made that 6.0 and not 5.9 if you're following SemVer.

**Comment by ENikS** [comment]
I had my reasons to continue with version 5.9

Although Unity.Abstractions had a lot of changes, Unity.Container is just incremental adaptation of these changes. It did not fundamentally change to warrant major version release. It is still the same container as v5.8, just recompiled to be compliant with adjusted public spec.

Unity 6 will be major overhaul of the container's engine and will not be compatible with 5.9, although both will support new Unity.Abstractions v4.

**Comment by bording** [comment]
If you really had to change the strong naming key on Unity.Abstractions, then you should have kept 5.x using 3.3.0 instead.

**Comment by ENikS** [comment]
I will keep it in mind for the next time. 

I was debating if I should continue support for new features of v4 spec in v5 container or create clean cut with v6 (Abstractions v4 and Container v6). I've decided to allow both v5 and v6 containers to support v4 spec to provide easier transition from one technology to the next. It did create breaking change in v5 line though. 
So, if you are so upset about this decision you have a choice, stay on spec v3.3 and container v5.8 till v6 is out. Next update will be as breaking but will follow SemVer to the letter.

**Comment by timbussmann** [comment]
closing this as this is "fixed" by 9.0.1 and 8.1.4 which restrict the dependency range. There will be a new version soon which will support unity 5.9.

Thanks for raising this @danrien 👍
